### PR TITLE
MINIFICPP-1121 - Upgrade spdlog to version 1.8.0

### PR DESCRIPTION
**Pull request by hunyadi-dev**
This is a draft PR used for running CI jobs on different platforms.

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/include/core/PropertyValidation.h
Comment (fgerlits) [inline libminifi/include/core/PropertyValidation.h:205]: the `()` around `max` can be removed, too

File: libminifi/src/utils/file/FileUtils.cpp
Comment (fgerlits) [inline libminifi/src/utils/file/FileUtils.cpp:42]: this comment is no longer needed

File: libminifi/include/core/PropertyValidation.h
Comment (hunyadi-dev) [reply libminifi/include/core/PropertyValidation.h]: Good catch, removed.

File: libminifi/src/utils/file/FileUtils.cpp
Comment (hunyadi-dev) [reply libminifi/src/utils/file/FileUtils.cpp]: Good catch, removed.

File: LICENSE
Comment (szaszm) [inline LICENSE:281]: Could you also 
1. update the copyright notices and 
2. include the fmt license text here?

Comment (szaszm) [reply LICENSE:281]: And please update our NOTICE file to include spdlog, fmt and their respective copyright notices.

Comment (hunyadi-dev) [reply LICENSE:281]: I am not exactly sure how the first comments first bullet differs from the second comment. Updated the `NOTICE` and added a section for fmt in this `LICENSE` file.

Comment (szaszm) [reply LICENSE:281]: In the `LICENSE` file lines 248-260, there is an outdated list of copyright notices. Please update those to reflect the current version of spdlog's [`LICENSE`](https://github.com/gabime/spdlog/blob/v1.7.0/LICENSE#L3) file

File: NOTICE
Comment (szaszm) [inline NOTICE:53]: The dot was part of the original: https://github.com/gsl-lite/gsl-lite/blob/master/LICENSE#L4

File: LICENSE
Comment (hunyadi-dev) [reply LICENSE:277]: Ah, I have seen those lines, but I do not know where they are coming from. At no point did spdlog have them listed in its `LICENSE` ([[see here]](https://github.com/gabime/spdlog/commits/v1.x/LICENSE)), but their names were present in the codebase as copyright holders. For some reason [[Aldrin asked us]](https://github.com/apache/nifi-minifi-cpp/pull/55) to provide these in our notice as well.

I updated the lines so that they are in correspondence with the current copyright notices in the spdlog files.

File: NOTICE
Comment (hunyadi-dev) [reply NOTICE:53]: Ah, I did not know that these notices are literal. Will add it back again :)

Comment (szaszm) [reply NOTICE:53]: I don't know whether they need to be literal, but I tried to modify as little as possible while keeping then on a single line (for 1 3rd party)

File: LICENSE
Comment (szaszm) [reply LICENSE:277]: Good point from Aldrin, I take back that point. It's too bad that spdlog's LICENSE file is incomplete... :(

File: CMakeLists.txt
Comment (fgerlits) [inline CMakeLists.txt:271]: this could go to the top level, around line 111, as `add_definitions(-DNOMINMAX)`, I think

Comment (hunyadi-dev) [reply CMakeLists.txt]: They moved up there :)

File: libminifi/src/core/logging/LoggerConfiguration.cpp
Comment (szaszm) [reply libminifi/src/core/logging/LoggerConfiguration.cpp:302]:
```suggestion
  return spdlog::syslog_logger_mt("ApacheNiFiMiNiFi", 0, LOG_USER, false);
```

File: libminifi/src/core/logging/LoggerConfiguration.cpp
Comment (szaszm) [inline libminifi/src/core/logging/LoggerConfiguration.cpp:318]: Instead of the 3 lines above:
```
result->sinks = { std::make_shared<spdlog::sinks::stderr_sink_mt>(); };
```

File: libminifi/src/core/logging/LoggerConfiguration.cpp
Comment (hunyadi-dev) [reply libminifi/src/core/logging/LoggerConfiguration.cpp:302]: I tried doing the suggested change before updating the PR, but unfortunately it does not compile.

File: libminifi/src/core/logging/LoggerConfiguration.cpp
Comment (hunyadi-dev) [reply libminifi/src/core/logging/LoggerConfiguration.cpp:318]: That is a valid and useful refactor. I did not check the context on doing the update.

**Comment by szaszm** [comment]
Could you rebase to main? Two reasons: 1. there are conflicts, 2. there are more CI failures than the usual flickers, but they don't seem to be related, so I would like to see another run.

**Comment by hunyadi-dev** [comment]
The slow speed on the builds are probably not only due to cache misses. Local builds (on a MacBook Pro 16", 2019) show similar clean build speeds when not using ccache:
```bash
cd .. && rm -rf build && mkdir build && cd build && cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug -DFORCE_COLORED_OUTPUT=ON -DENABLE_LIBRDKAFKA=ON -DENABLE_PYTHON=ON -DENABLE_COAP=ON -DSKIP_TESTS=  -DUSE_SHARED_LIBS=ON  -DPORTABLE=ON  -DBUILD_ROCKSDB=ON  -DBUILD_IDENTIFIER= ../src && export CCACHE_DISABLE=1 && time ninja -j4
```
1. **On this branch:**

    > ninja -j4  3536.57s user 429.61s system 442% cpu 14:56.78 total
1. **Reference (main):**

    > ninja -j4  2823.80s user 439.10s system 457% cpu 11:53.96 total

Please review for potential loss of build speeds. Is it worth the effort of upgrading?

File: cmake/BundledSpdlog.cmake
Comment (szaszm) [inline cmake/BundledSpdlog.cmake:46]: Would you mind trying to bump this to 1.8.0?

Comment (hunyadi-dev) [reply cmake/BundledSpdlog.cmake]: Updated.

**Comment by szaszm** [comment]
I got rid of the cmake issue in `make debian`.
- by upgrading to debian 10 buster: https://github.com/szaszm/nifi-minifi-cpp/tree/spdlog_upgrade_debian_buster
- by using stretch-backports and stretch-backports-sloppy (due to an accidental upload to backports by the package maintainer): https://github.com/szaszm/nifi-minifi-cpp/tree/MINIFICPP-1121_spdlog_upgrade

**Comment by hunyadi-dev** [comment]
> I got rid of the cmake issue in `make debian`.
> 
>     * by upgrading to debian 10 buster: [https://github.com/szaszm/nifi-minifi-cpp/tree/spdlog_upgrade_debian_buster](https://github.com/szaszm/nifi-minifi-cpp/tree/spdlog_upgrade_debian_buster?rgh-link-date=2020-11-20T12%3A42%3A17Z)
> 
>     * by using stretch-backports and stretch-backports-sloppy (due to an accidental upload to backports by the package maintainer): [https://github.com/szaszm/nifi-minifi-cpp/tree/MINIFICPP-1121_spdlog_upgrade](https://github.com/szaszm/nifi-minifi-cpp/tree/MINIFICPP-1121_spdlog_upgrade?rgh-link-date=2020-11-20T12%3A42%3A17Z)

Thanks, that is a great addition, I cherry picked the latter to this PR.

**Review by szaszm — APPROVED** [review]
I think this is ready for merge but let's wait for one approval besides mine because of the latest changes. @fgerlits @arpadboda

File: cmake/BundledSpdlog.cmake
Comment (fgerlits) [reply cmake/BundledSpdlog.cmake:32]: What happened here?  I think I remember a commit which applied these changes, but now the generator expressions are not there.

File: docker/debian/Dockerfile
Comment (fgerlits) [inline docker/debian/Dockerfile:34]: Why do we need the system libarchive?  We bundle that.

**Review by fgerlits — COMMENTED** [review]
Minor, but the PR comment still says this is a draft PR.

File: cmake/BundledSpdlog.cmake
Comment (fgerlits) [inline cmake/BundledSpdlog.cmake:32]: What happened here? I think I remember a commit which changed these to generator expressions, but now the generator expressions are not there.

File: docker/debian/Dockerfile
Comment (szaszm) [reply docker/debian/Dockerfile:34]: It's a dependency of cmake. The maintainer incorrectly uploaded cmake 3.16 (instead of 3.13) to the amd64 stretch-backports, which depends on libarchive >=3.3.3. The version in the stretch repos is 3.2 and there is no libarchive in stretch-backports, so this is a workaround. 

See this bug: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=954852
and see the inconsistent version near the bottom of this page: https://packages.debian.org/stretch-backports/cmake

File: cmake/BundledSpdlog.cmake
Comment (szaszm) [reply cmake/BundledSpdlog.cmake:32]: https://gitlab.kitware.com/cmake/cmake/-/issues/19522

**Review by fgerlits — APPROVED** [review]

File: cmake/BundledSpdlog.cmake
Comment (adamdebreceni) [inline cmake/BundledSpdlog.cmake:72]: to prevent increasing the compile time

```suggestion
        set_property(TARGET spdlog APPEND PROPERTY INTERFACE_COMPILE_DEFINITIONS "SPDLOG_ENABLE_SYSLOG" "SPDLOG_COMPILED_LIB")
```

**Review by adamdebreceni — CHANGES_REQUESTED** [review]

**Comment by szaszm** [comment]
There are some compilation errors caused by the last change

**Comment by hunyadi-dev** [comment]
@szaszm rebased and updated again...

**Comment by szaszm** [comment]
This is the issue: https://github.com/apache/nifi-minifi-cpp/pull/837/checks?check_run_id=1522646567#step:5:14323
Similar to ec40383, probably just missing includes.

**Review by adamdebreceni — APPROVED** [review]

**Pull request closed without merging**
